File: projects/rpn2tex/migrations/java-module-by-module-2/src/test/java/com/rpn2tex/TokenTest.java
package com.rpn2tex;

import org.junit.jupiter.api.Test;
import org.junit.jupiter.params.ParameterizedTest;
import org.junit.jupiter.params.provider.CsvSource;
import org.junit.jupiter.params.provider.ValueSource;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertThrows;

/**
 * Unit tests for the {@link Token} record.
 *
 * <p>Tests cover:
 * <ul>
 *   <li>Token creation with valid parameters</li>
 *   <li>Immutability guarantees</li>
 *   <li>String representation formatting</li>
 *   <li>Position information (1-based indexing)</li>
 *   <li>Validation of null and invalid parameters</li>
 * </ul>
 */
class TokenTest {

    @Test
    void testTokenCreation() {
        Token token = new Token(TokenType.NUMBER, "42", 1, 5);

        assertNotNull(token);
        assertEquals(TokenType.NUMBER, token.type());
        assertEquals("42", token.value());
        assertEquals(1, token.line());
        assertEquals(5, token.column());
    }

    @ParameterizedTest
    @CsvSource({
        "NUMBER, '5', 1, 1",
        "NUMBER, '3.14', 1, 5",
        "PLUS, '+', 1, 3",
        "MINUS, '-', 2, 10",
        "MULT, '*', 3, 7",
        "DIV, '/', 1, 9",
        "EOF, '', 5, 20"
    })
    void testTokenCreationWithVariousValues(String typeStr, String value, int line, int column) {
        TokenType type = TokenType.valueOf(typeStr);
        Token token = new Token(type, value, line, column);

        assertEquals(type, token.type());
        assertEquals(value, token.value());
        assertEquals(line, token.line());
        assertEquals(column, token.column());
    }

    @Test
    void testToStringFormat() {
        Token token = new Token(TokenType.NUMBER, "42", 1, 5);
        String result = token.toString();

        // Expected format: Token(TYPE, 'value', line:column)
        assertEquals("Token(NUMBER, '42', 1:5)", result);
    }

    @ParameterizedTest
    @CsvSource({
        "NUMBER, '5', 1, 1, 'Token(NUMBER, ''5'', 1:1)'",
        "PLUS, '+', 2, 3, 'Token(PLUS, ''+'', 2:3)'",
        "EOF, '', 10, 25, 'Token(EOF, '''', 10:25)'"
    })
    void testToStringFormatVariousTokens(String typeStr, String value, int line, int column,
                                          String expected) {
        TokenType type = TokenType.valueOf(typeStr);
        Token token = new Token(type, value, line, column);

        assertEquals(expected, token.toString());
    }

    @Test
    void testNullTypeThrowsException() {
        assertThrows(NullPointerException.class, () -> {
            new Token(null, "42", 1, 5);
        });
    }

    @Test
    void testNullValueThrowsException() {
        assertThrows(NullPointerException.class, () -> {
            new Token(TokenType.NUMBER, null, 1, 5);
        });
    }

    @ParameterizedTest
    @ValueSource(ints = {0, -1, -10, Integer.MIN_VALUE})
    void testInvalidLineNumberThrowsException(int invalidLine) {
        assertThrows(IllegalArgumentException.class, () -> {
            new Token(TokenType.NUMBER, "42", invalidLine, 5);
        });
    }

    @ParameterizedTest
    @ValueSource(ints = {0, -1, -10, Integer.MIN_VALUE})
    void testInvalidColumnNumberThrowsException(int invalidColumn) {
        assertThrows(IllegalArgumentException.class, () -> {
            new Token(TokenType.NUMBER, "42", 1, invalidColumn);
        });
    }

    @Test
    void testMinimumValidPosition() {
        // Line 1, Column 1 is the minimum valid position
        Token token = new Token(TokenType.NUMBER, "42", 1, 1);

        assertEquals(1, token.line());
        assertEquals(1, token.column());
    }

    @Test
    void testLargeLineAndColumnNumbers() {
        // Test with large but valid position numbers
        Token token = new Token(TokenType.NUMBER, "42", 1000, 5000);

        assertEquals(1000, token.line());
        assertEquals(5000, token.column());
    }

    @Test
    void testRecordEquality() {
        // Records automatically implement equals() based on all fields
        Token token1 = new Token(TokenType.NUMBER, "42", 1, 5);
        Token token2 = new Token(TokenType.NUMBER, "42", 1, 5);
        Token token3 = new Token(TokenType.NUMBER, "43", 1, 5);

        assertEquals(token1, token2, "Identical tokens should be equal");
        assertEquals(token1.hashCode(), token2.hashCode(),
                     "Identical tokens should have same hash code");

        // Different value should make tokens unequal
        assertEquals(false, token1.equals(token3),
                     "Tokens with different values should not be equal");
    }

    @Test
    void testRecordImmutability() {
        // Records are immutable - this test verifies the accessor methods work
        Token token = new Token(TokenType.PLUS, "+", 3, 7);

        // Multiple calls should return the same values
        assertEquals(TokenType.PLUS, token.type());
        assertEquals(TokenType.PLUS, token.type());
        assertEquals("+", token.value());
        assertEquals("+", token.value());
        assertEquals(3, token.line());
        assertEquals(3, token.line());
        assertEquals(7, token.column());
        assertEquals(7, token.column());
    }

    @Test
    void testEmptyValueAllowed() {
        // EOF tokens often have empty values
        Token token = new Token(TokenType.EOF, "", 1, 10);

        assertEquals("", token.value());
        assertEquals("Token(EOF, '', 1:10)", token.toString());
    }

    @Test
    void testNumberTokenWithDecimal() {
        Token token = new Token(TokenType.NUMBER, "3.14", 1, 1);

        assertEquals(TokenType.NUMBER, token.type());
        assertEquals("3.14", token.value());
    }

    @Test
    void testNumberTokenWithNegative() {
        Token token = new Token(TokenType.NUMBER, "-42", 1, 1);

        assertEquals(TokenType.NUMBER, token.type());
        assertEquals("-42", token.value());
    }

    @Test
    void testOperatorTokens() {
        // Test all operator types
        Token plusToken = new Token(TokenType.PLUS, "+", 1, 1);
        Token minusToken = new Token(TokenType.MINUS, "-", 1, 2);
        Token multToken = new Token(TokenType.MULT, "*", 1, 3);
        Token divToken = new Token(TokenType.DIV, "/", 1, 4);

        assertEquals("+", plusToken.value());
        assertEquals("-", minusToken.value());
        assertEquals("*", multToken.value());
        assertEquals("/", divToken.value());
    }
}

File: projects/rpn2tex/migrations/go-module-by-module-2/token_test.go
package rpn2tex

import (
	"testing"
)

func TestTokenTypeString(t *testing.T) {
	tests := []struct {
		name     string
		tokType  TokenType
		expected string
	}{
		{"NUMBER", NUMBER, "NUMBER"},
		{"PLUS", PLUS, "PLUS"},
		{"MINUS", MINUS, "MINUS"},
		{"MULT", MULT, "MULT"},
		{"DIV", DIV, "DIV"},
		{"EOF", EOF, "EOF"},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got := tt.tokType.String()
			if got != tt.expected {
				t.Errorf("TokenType.String() = %q, want %q", got, tt.expected)
			}
		})
	}
}

func TestTokenCreation(t *testing.T) {
	tests := []struct {
		name       string
		token      Token
		wantType   TokenType
		wantValue  string
		wantLine   int
		wantColumn int
	}{
		{
			name: "NUMBER token",
			token: Token{
				Type:   NUMBER,
				Value:  "42",
				Line:   1,
				Column: 5,
			},
			wantType:   NUMBER,
			wantValue:  "42",
			wantLine:   1,
			wantColumn: 5,
		},
		{
			name: "PLUS token",
			token: Token{
				Type:   PLUS,
				Value:  "+",
				Line:   2,
				Column: 3,
			},
			wantType:   PLUS,
			wantValue:  "+",
			wantLine:   2,
			wantColumn: 3,
		},
		{
			name: "EOF token",
			token: Token{
				Type:   EOF,
				Value:  "",
				Line:   3,
				Column: 10,
			},
			wantType:   EOF,
			wantValue:  "",
			wantLine:   3,
			wantColumn: 10,
		},
		{
			name: "Decimal number token",
			token: Token{
				Type:   NUMBER,
				Value:  "3.14",
				Line:   1,
				Column: 1,
			},
			wantType:   NUMBER,
			wantValue:  "3.14",
			wantLine:   1,
			wantColumn: 1,
		},
		{
			name: "Negative number token",
			token: Token{
				Type:   NUMBER,
				Value:  "-5",
				Line:   1,
				Column: 1,
			},
			wantType:   NUMBER,
			wantValue:  "-5",
			wantLine:   1,
			wantColumn: 1,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			if tt.token.Type != tt.wantType {
				t.Errorf("Token.Type = %v, want %v", tt.token.Type, tt.wantType)
			}
			if tt.token.Value != tt.wantValue {
				t.Errorf("Token.Value = %q, want %q", tt.token.Value, tt.wantValue)
			}
			if tt.token.Line != tt.wantLine {
				t.Errorf("Token.Line = %d, want %d", tt.token.Line, tt.wantLine)
			}
			if tt.token.Column != tt.wantColumn {
				t.Errorf("Token.Column = %d, want %d", tt.token.Column, tt.wantColumn)
			}
		})
	}
}

func TestTokenString(t *testing.T) {
	tests := []struct {
		name     string
		token    Token
		expected string
	}{
		{
			name: "NUMBER token string",
			token: Token{
				Type:   NUMBER,
				Value:  "42",
				Line:   1,
				Column: 5,
			},
			expected: `Token(NUMBER, "42", 1:5)`,
		},
		{
			name: "PLUS token string",
			token: Token{
				Type:   PLUS,
				Value:  "+",
				Line:   2,
				Column: 3,
			},
			expected: `Token(PLUS, "+", 2:3)`,
		},
		{
			name: "MINUS token string",
			token: Token{
				Type:   MINUS,
				Value:  "-",
				Line:   1,
				Column: 7,
			},
			expected: `Token(MINUS, "-", 1:7)`,
		},
		{
			name: "MULT token string",
			token: Token{
				Type:   MULT,
				Value:  "*",
				Line:   1,
				Column: 9,
			},
			expected: `Token(MULT, "*", 1:9)`,
		},
		{
			name: "DIV token string",
			token: Token{
				Type:   DIV,
				Value:  "/",
				Line:   1,
				Column: 11,
			},
			expected: `Token(DIV, "/", 1:11)`,
		},
		{
			name: "EOF token string",
			token: Token{
				Type:   EOF,
				Value:  "",
				Line:   3,
				Column: 10,
			},
			expected: `Token(EOF, "", 3:10)`,
		},
		{
			name: "Decimal number token string",
			token: Token{
				Type:   NUMBER,
				Value:  "3.14",
				Line:   1,
				Column: 1,
			},
			expected: `Token(NUMBER, "3.14", 1:1)`,
		},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got := tt.token.String()
			if got != tt.expected {
				t.Errorf("Token.String() = %q, want %q", got, tt.expected)
			}
		})
	}
}

func TestTokenTypeComparison(t *testing.T) {
	tests := []struct {
		name  string
		tok1  TokenType
		tok2  TokenType
		equal bool
	}{
		{"NUMBER equals NUMBER", NUMBER, NUMBER, true},
		{"PLUS equals PLUS", PLUS, PLUS, true},
		{"NUMBER not equals PLUS", NUMBER, PLUS, false},
		{"MULT not equals DIV", MULT, DIV, false},
		{"EOF equals EOF", EOF, EOF, true},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got := tt.tok1 == tt.tok2
			if got != tt.equal {
				t.Errorf("(%v == %v) = %v, want %v", tt.tok1, tt.tok2, got, tt.equal)
			}
		})
	}
}

func TestTokenPositionTracking(t *testing.T) {
	// Verify that tokens correctly preserve their position information
	// This is critical for error reporting
	token := Token{
		Type:   NUMBER,
		Value:  "123",
		Line:   5,
		Column: 12,
	}

	if token.Line != 5 {
		t.Errorf("Position tracking failed: Line = %d, want 5", token.Line)
	}
	if token.Column != 12 {
		t.Errorf("Position tracking failed: Column = %d, want 12", token.Column)
	}
}

func TestAllOperatorTypes(t *testing.T) {
	// Test that all operator types are correctly represented
	operators := []struct {
		tokType TokenType
		value   string
		name    string
	}{
		{PLUS, "+", "PLUS"},
		{MINUS, "-", "MINUS"},
		{MULT, "*", "MULT"},
		{DIV, "/", "DIV"},
	}

	for _, op := range operators {
		t.Run(op.name, func(t *testing.T) {
			token := Token{
				Type:   op.tokType,
				Value:  op.value,
				Line:   1,
				Column: 1,
			}

			if token.Type != op.tokType {
				t.Errorf("Expected type %v, got %v", op.tokType, token.Type)
			}
			if token.Value != op.value {
				t.Errorf("Expected value %q, got %q", op.value, token.Value)
			}
			if token.Type.String() != op.name {
				t.Errorf("Expected name %q, got %q", op.name, token.Type.String())
			}
		})
	}
}
